### DISCUSSION TITLE
List view improvements

### DISCUSSION THREAD

**Discussion by philipdanielhayton** [umbraco/Umbraco-CMS#16628] [Features and ideas]
Hi guys,

I'm just playing around with v14 of Umbraco and was surprised to see that the behavior of the list view has changed. Specifically, doc types set as a list view will still show their child items in the tree view.

Anyway, that got me thinking, if there are changes being made to the list view, maybe this is a good opportunity to improve and expand it's functionality?

Two things that I've always longed for are:

- The ability to only show certain doc types in the list, but continue to show others in the tree. A good use case for this is a blog section where you will typically have a handful of categories which could be shown in the tree, but thousands of articles that could be shown in the list.
- Additionally and / or alternatively, it would be great to be able to have more than one list view per doc type. 
- Oh, and if we could at least have the option to hide list view items from the tree, that would be great! 😁

Anyway, this is mostly just me thinking out loud, would be interested to hear what others think? How would you improve it?

**Comment by skttl** [comment]
I think it's a bug that list views are showing children in the tree, but other than that, I would love to have those features too!

**Reply by nielslyngsoe** [reply]
Hi @skttl and others, its not intended as a bug.

The background story is that the List View was made for a more detailed overview and to solve a performance problem.

We do not have the performance problem of the tree any longer, as we have implemented pagination, aka. load more.
In this way the List View is not seen as a way to hide children from the tree. But as a detail overview.

This enables users to still navigate via the tree, if they like. As well it provides an element of confirmation that you can see that there is child items within, not hidden inside the list view. My perspective here is mainly towards new users, those who do not know the List View concept.

It may still be relevant to be able to hide children in the tree, but I would like to hear and see the use cases before we invest time and effort into such feature. So please share more concrete examples of such.

Thanks in advance

**Reply by PeterKvayt** [reply]
@nielslyngsoe 
So we have the following case: old umbraco users expect that there is no content in the tree when we using list view, new umbraco users do not expect any.

I think instead of choose show content in tree or not, let's provide ability to enable content in tree or not in options of list view.

What do you think about it?

**Reply by skttl** [reply]
One case of needing to hide the children in the tree could be when having a different sort order. Eg. by article date or similar. In this case it would just confuse the editors that the articles get listed in another order in the tree.

I dream of being able to select which content types go into the list view, and which goes into the tree. So on a specific list view, you can specify that the list view should only contain nodes of one or more specific content types, and additionally configure these content types to be hidden from the tree.

**Reply by dennispedersenakqa** [reply]
@nielslyngsoe Is this still the way you are looking at things. We might have hit another issue with it.

We are experiencing very, very long login times with a solution that has alot of migrated pages, and we suspect its because its loading the entire tree. however we cannot be sure, as its the login action that takes time. We have a login times of over 2 hours and counting, when we have imported around 50k nodes in umbraco. This is not really useable. Is this something you've experienced in your tests?
This is a huge case for us, and we dont really know what to do, since we cant really circumvent the login in any way. 

I think for the best user scenario would be have a the ability to hide the tree nodes when making list views.

**Reply by netaddicts-council** [reply]
> It may still be relevant to be able to hide children in the tree, but I would like to hear and see the use cases before we invest time and effort into such feature. So please share more concrete examples of such.

It's completely relevant, and the discussion is already open for a while now, and honestly, not sure why it didn't get addressed... Our system is having ~2k press releases sitting in the same section, and it just doesn't make any sense to see any items in the navigation tree at all. Again, you can always argue why so many content items living under the same parent (and not having a folder system), but that's just how it works now and has for the last decade (ok, little exaggeration here), and sometimes it's also pure legacy
A "simple" extra configuration on the listview should do the trick. If set, do not expand, it not, use whatever is used now (I'm not a huge fan of this pagination, but also don't have big claims to remove it)

I also do like suggestion of having a filtered tree (only for listviews), cause it's indeed a common scenario to have multiple doc types configured living under the same parent (Although I would believe that implementation is far less obvious

**Reply by tomotrepo** [reply]
I agree with all the comments. We have a similar situation, where we have a 'content part' that is branched throughout the tree structure and we have a 'catalogue part', such as news, events, accommodations, etc. in tabelar structure. For example, accommodations do not need (or should not have) a branched structure, since everything is filtered via properties: stars, shower, wifi, parking etc. And the tabelar system wins here - for years. But the biggest problem is our editors. They have absolutely no idea what is going on. Now they are editing one accommodation in the table and tree on the left it keeps showing on and off some accommodations which are irrelevant for editor!? How should I explain to editor that it is now there just because now it can be?
We are not against it, although to this day I have not found a reason why it would be useful to me, we would suggest that there is a setting in appsetting.json whether we want to have it or not, let everyone decide for themselves. Old-school Umbraco administrators will just turn it off, because it brings more headaches than benefits.

Regards, Tomo

**Reply by nielslyngsoe** [reply]
I also agree with all the comments, and we surely need to do something about it.

But please be helpful and give us a better basis for the next step. So please share more concrete examples — Michael's comment further below is super useful.

The general problem with configuration options is that it is not helpful for newcomers, and it eliminates our options to improve and secure a good out-of-the-box experience.

In other words: The more granular the configuration — the harder the lock-in on a config-specific UX. This gives less room to evolve within and makes it more complex for newcomers to make the right choices.

Let's make the experience great for everyone.

**Comment by mistyn8** [comment]
Bit late to the party (and have to admit not made the jump to v14/14 as yet)

Maybe not a concrete example for hiding listView items in the tree but if this is also a wider remove the listview as the problem it solved is no longer an issue... 

then list views have uses beyond that!

- list and grid view with custom props exposed
- bulk actions on the list
- scoped search within a list view. (be great to revisit searching custom props too)

**Reply by hfloyd** [reply]
I agree that listviews have lots of uses, like those mentioned by @mistyn8 (in pre-v14 site I have often added listviews as properties on doctypes when I wanted these features without hiding the tree items). If conversations are happening about the future of listviews, may I also suggest these other discussions:
https://github.com/umbraco/Umbraco-CMS/discussions/13295
https://github.com/umbraco/Umbraco-CMS/discussions/12393

**Comment by ssplyuvrajsingh** [comment]
Hi @nielslyngsoe ,
I am using list view in my umbraco 14 project which I have upgraded from  umbrco 13 to 14 but I am unable to add custom search functionality in umbraco 14 which i was using in umbraco 13 ,I Implemented to see this [CLICK TO SEE](https://dev.to/skttl/how-to-customize-searching-in-umbraco-list-views-1knk#:~:text=Using%20the%20magic%20of%20$http,out%2C%20with%20your%20own%20endpoint.)

**My js fie**

```
angular.module('umbraco.services').config([
    '$httpProvider',
    function ($httpProvider) {
        $httpProvider.interceptors.push(function ($q) {
            return {
                'request': function (request) {
                    // Redirect any requests for the listview to our custom list view UI
                    if (request.url.indexOf("umbraco/backoffice/umbracoapi/content/GetChildren?id=") > -1)
                    request.url = request.url.replace("backoffice/umbracoapi/content/GetChildren", "backoffice/Umbracoapi/CustomListViewSearch/GetChildrenCustom");

                    return request || $q.when(request);
                }
            };
        });

    }]);
```
**An my Controller**

```
using Examine;
using Examine.Search;
using Microsoft.AspNetCore.Authorization;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using NPoco;
using Umbraco.Cms.Core;
using Umbraco.Cms.Core.Actions;
using Umbraco.Cms.Core.Configuration.Models;
using Umbraco.Cms.Core.Dictionary;
using Umbraco.Cms.Core.Events;
using Umbraco.Cms.Core.Mapping;
using Umbraco.Cms.Core.Models;
using Umbraco.Cms.Core.Models.ContentEditing;
using Umbraco.Cms.Core.Persistence.Querying;
using Umbraco.Cms.Core.PropertyEditors;
using Umbraco.Cms.Core.Routing;
using Umbraco.Cms.Core.Scoping;
using Umbraco.Cms.Core.Security;
using Umbraco.Cms.Core.Serialization;
using Umbraco.Cms.Core.Services;
using Umbraco.Cms.Core.Strings;
using Umbraco.Cms.Infrastructure.Persistence;
using Umbraco.Cms.Web.BackOffice.Controllers;
using Umbraco.Cms.Web.Common.Attributes;
using Umbraco.Extensions;

namespace Rex.Repository.Controllers.Plugin
{
    [PluginController("UmbracoApi")]
    public class CustomListViewSearchController : ContentController
    {
        private readonly IContentService _contentService;
        //private readonly ILogger _logger;
        private readonly IExamineManager _examineManager;
        private readonly IUmbracoMapper _mapper;
        private readonly Lazy<IDictionary<string, ILanguage>> _allLangs;
        private readonly ILocalizationService _localizationService;
        private readonly ISqlContext _sqlContext;
        public CustomListViewSearchController(ICultureDictionary cultureDictionary,
         ILoggerFactory loggerFactory,
         IShortStringHelper shortStringHelper,
         IEventMessagesFactory eventMessages,
         ILocalizedTextService localizedTextService,
         PropertyEditorCollection propertyEditors,
         IContentService contentService,
         IUserService userService,
         IBackOfficeSecurityAccessor backofficeSecurityAccessor,
         IContentTypeService contentTypeService,
         IUmbracoMapper umbracoMapper,
         IPublishedUrlProvider publishedUrlProvider,
         IDomainService domainService,
         IDataTypeService dataTypeService,
         ILocalizationService localizationService,
         IFileService fileService,
         INotificationService notificationService,
         ActionCollection actionCollection,
         ISqlContext sqlContext,
         IJsonSerializer serializer,
         ICoreScopeProvider scopeProvider,
         IAuthorizationService authorizationService,
         IContentVersionService contentVersionService,
         ICultureImpactFactory cultureImpactFactory,
         IOptions<ContentSettings> contentSettings, IExamineManager examineManager, IUmbracoMapper mapper)
             : base(cultureDictionary, loggerFactory, shortStringHelper, eventMessages, localizedTextService, propertyEditors, contentService, userService, backofficeSecurityAccessor, contentTypeService, umbracoMapper, publishedUrlProvider, domainService, dataTypeService, localizationService, fileService, notificationService, actionCollection, sqlContext, serializer, scopeProvider, authorizationService, contentVersionService, cultureImpactFactory, contentSettings)
       {
            _contentService = contentService;
            _examineManager = examineManager;
            _mapper = mapper;
            _localizationService = localizationService;
            _allLangs = new Lazy<IDictionary<string, ILanguage>>(() =>
        _localizationService.GetAllLanguages().ToDictionary(x => x.IsoCode, x => x, StringComparer.InvariantCultureIgnoreCase));
            _sqlContext= sqlContext;
        }

        public PagedResult<ContentItemBasic<ContentPropertyBasic>> GetChildrenCustom(int id, string includeProperties, int pageNumber = 0, int pageSize = 0, string orderBy = "SortOrder", Direction orderDirection = Direction.Ascending, bool orderBySystemField = true, string filter = "", string cultureName = "")
        {
            // get the parent node, and its doctype alias from the content cache
            var parentNode = _contentService.GetById(id);
            var parentNodeDocTypeAlias = parentNode != null ? parentNode.ContentType.Alias : null;

            // if the parent node is not "books", redirect to the core GetChildren() method
            if (parentNode?.ContentType.Alias != "products" || filter.IsNullOrWhiteSpace())
            {
                return GetChildren(id, includeProperties, pageNumber, pageSize, orderBy, orderDirection, orderBySystemField, filter, cultureName);
            }

            // if we can't get the InternalIndex, redirect to the core GetChildren() method, but log an error
            if (!_examineManager.TryGetIndex("ProductIndex", out IIndex index))
            {
                //_logger.Error<CustomListViewSearchController>("Couldn't get InternalIndex for searching products in list view");
                return GetChildren(id, includeProperties, pageNumber, pageSize, orderBy, orderDirection, orderBySystemField, filter, cultureName);
            }

            // find children using Examine

            // create search criteria
            var searcher = index.Searcher; 
            var searchCriteria = searcher.CreateQuery();
            var searchQuery = searchCriteria.Field("parentID",Convert.ToString(id));

            if (!filter.IsNullOrWhiteSpace())
            {
                searchQuery = searchQuery.And().ManagedQuery(filter,new[] { "nodeName", "SKU" });
            }

           
            // pageNumber is not zero indexed in this, so just multiply pageSize by pageNumber
            var searchResults = searchQuery.Execute();

            // get the results on the current page
            // pageNumber is not zero indexed in this, so subtract 1 from the pageNumber
            var totalChildren = searchResults.TotalItemCount;

            if (totalChildren == 0)
            {
                return GetChildren(id, includeProperties, pageNumber, pageSize, orderBy, orderDirection, orderBySystemField, filter, cultureName);
            }

			var pagedResultIds = searchResults.Skip((pageNumber > 0 ? pageNumber - 1 : 0) * pageSize).Take(pageSize).Select(x => x.Id).Select(x => int.Parse(x)).ToList();
            var children = _contentService.GetByIds(pagedResultIds).ToList();

            var pagedResult = new PagedResult<ContentItemBasic<ContentPropertyBasic>>(totalChildren, pageNumber, pageSize)
            {
                Items = children.Select(content =>
                    _mapper.Map<IContent, ContentItemBasic<ContentPropertyBasic>>(
                        content,
                        context =>
                        {
                            context.SetCulture(cultureName);

                            // if there's a list of property aliases to map - we will make sure to store this in the mapping context.
                            if (!includeProperties.IsNullOrWhiteSpace())
                            {
                                context.SetIncludedProperties(includeProperties.Split(
                                    new[] { ", ", "," },
                                    StringSplitOptions.RemoveEmptyEntries));
                            }
                        }))
           .WhereNotNull()
           .ToList() // evaluate now
            };

            return pagedResult;
        }
    }
}
```

**Please provide me a way so I can implement it to my umbraco 14 project**

**Reply by hfloyd** [reply]
@ssplyuvrajsingh - I'd suggest moving this specific message to the [Forum ](https://forum.umbraco.com/)where you are more likely to get some help with your code.

**Comment by Myster** [comment]
I separately posted a feature request to [make the listview available on any page](https://github.com/umbraco/Umbraco-CMS/discussions/19272). This idea is different/tangential to the ideas discussed here.

**Comment by tomotrepo** [comment]
Hi everyone,

I would like to add my vote to this and request that this issue be reconsidered.
We have been using Umbraco extensively since version 7. After recently upgrading our projects to v17, we are facing a lot of pushback from our content editors regarding the new Collection behavior.
Because the child nodes now appear in the content tree alongside the list view, many of our editors are completely confused. They perceive this as a bug and are constantly asking our development team to "fix" or remove it. We try our best to explain that this is now "by design" due to the new pagination features, but honestly, many of them just don't understand the reasoning and find the content tree incredibly cluttered and annoying to navigate—especially on nodes that contain thousands of articles or news items.
While we understand the technical improvements behind the new tree, the user experience for long-time Umbraco editors has taken a step backward here.
I strongly vote to reopen this discussion and kindly ask the HQ to consider adding at least a simple toggle button (e.g., "Hide children from tree") in the Collection configuration. This would give developers the flexibility to choose the best approach for their specific use cases and keep the editors happy.

Thank you!

**Comment by bjarnef** [comment]
Recently I noticed we have the option to adjust "Workspace View name", but create button always seems to use name of content type.
It would be useful to override this.

There are other improvements in the visual part of how data is presented to content editors. One thing is better formatting of date time values depending on backoffice user locale:
https://github.com/umbraco/Umbraco-CMS/discussions?discussions_q=is%3Aopen+ufm

Another thing it dynamic filters on collection data, which I think @madsrasmussen has been worked on recently.
Besides the free text search: 

- Boolean value (toggle)
- Single value (radiobutton list)
- Multi values (checkbox list)
- Date (range) value (date picker)

Perhaps it should be possible to configure how the filter should work, e.g. if an int/bit value is handle as boolean and numeric value.

**Reply by bjarnef** [reply]
@nielslyngsoe would it be possible to override create button later via a config (or extension type)? Similar to workspace name and in block add button label.

Some developers may have element types / content types with same name, but therefore name these different. However for content editors seeing this label, it can often be simplified :)

**Comment by MichaelNielsenDK** [comment]
Nested list views are a clear case for being able to hide children from the tree. To me, and especially to editors, the ellipsis entries looks completely broken, like content has failed to load or an error has occurred.

Here is an example from a client project post v17 upgrade. 

<img width="296" height="726" alt="image" src="https://github.com/user-attachments/assets/c7be78fb-a767-4bb5-af1a-8971b3797c5f" />

**Reply by nielslyngsoe** [reply]
Hi @MichaelNielsenDK, interesting case that clearly pushes the boundaries; great💙

So in v.13 you would only have seen the path to `Afrika`; from there, the tree navigation would not provide any guidance to the user.

I'm not saying this is a great experience — clearly not — but if we were to provide just a bit of indication for the user of where they are. How would you suggest that could look?

**A suggestion from my side could be:**

Would it be fair if the other children on your path was hidden and the `...` was not present. So you only saw your direct path, without any representation of other children.

This would give the User an indication of where they are, and the ability to jump to a parent without being forced back to the first Collection (Afrika in your case).

So it would look like this:
<img width="296" height="354" alt="image" src="https://github.com/user-attachments/assets/0374d208-97d5-479e-b0bd-71d7541864c0" />

Let me hear what you think, cause your case clearly calls for action.

**Reply by nielslyngsoe** [reply]
I thought a little bit more on this problem; it would probably need to have some indication that prevents the user from thinking that the other children are missing.

Therefore, the solution might need to have some separations in play for that to be clear?
<img width="296" height="353" alt="image" src="https://github.com/user-attachments/assets/c7ee1759-8132-4e46-9a8e-60fe42b28c41" />

**Reply by mistyn8** [reply]
Not sure that I'm lost if the tree view doesn't show my current active node. Most file explorers work like that? (only if I actively expand the tree do I see the folder heirarchy)

(though maybe nice to have an indication that I'm in the tree hierarchy, so colour the ancestors to indicate that somehow.)

We have the breadcrumb path at the bottom of every content workspace (akin to the explorer path bar), so I'd be using that to navigate within a collection/nested collection space.

"So in v.13 you would only have seen the path to Afrika; from there, the tree navigation would not provide any guidance to the user." 
I personally don't have any problem with that.. as I'm fine with the breadcrumb path at that point, and shared experience with file managers that don't auto expand the tree :-)

I know windows11 explorer isn't an exemplar.. but directly going to a file.. has the three things I mention.. a breadcrumb style path, and indication of the ancestors and a tree that is usuable either collapsed by default, or able to be expanded with intent.
(and traversing in the file pane doesn't automatically expand/sync the folder tree)

<img width="973" height="335" alt="image" src="https://github.com/user-attachments/assets/e88f2a07-bf1a-4003-9002-48441da91632" />

**Reply by nielslyngsoe** [reply]
Fair @mistyn8, and thanks for pitching in.

In general, from a UX design perspective, people have a hard time with File systems, and Windows Explorer might hint at why.
— I´m not suspecting any developers having trouble with it, but the primary audience for this UX is Authors not devs.

In Umbraco, we expand the tree to show where you're at, very useful if someone sends you a link. So you get an overview of where you're at.

Yes, there is a breadcrumb, but it is really not ideal to pair that with the tree — also, many users never notice that there is a breadcrumb.

So the aim is to help authors understand where they are; the pink color helps make it easy to spot.

Michael's experience then makes it super clear that it is more of an annoyance and not any help, so great that he came forward with that insight. — So clear that something needs to be done, and the aim from my chair is to make both ends meet.

**Reply by mistyn8** [reply]
good response..  some feedback

>  So you get an overview of where you're at.

I'm fine, and my clients are too, that I'm shown I'm in the overarching top level collection..  (what they were used to in previous umbraco versions :-) )

> many users never notice that there is a breadcrumb

Should we acknowledge and address that and make it a first class citizen then, maybe in the header with the node name?

I think that's actually what you are trying to do with the current implementation, present the breadcrumb visually within the tree, when a breadcrumb would suffice?

Whilst @MichaelNielsenDK  suggestions improve on what we have.. I'd still say it's something I've never seen before.. a tree with pagination in it.. 
show previous.. show next.. 
still leaves a new user with no real idea of what clicking on it does.. 
 - previous what?
 - how many?
 - does it remove the current items and show the previous 10 like a pagination in a search results
 -  etc..
 
I think the rod you have created for you own back here is the insistence that the tree absolutely must show the current node, something that other trees don't insist as the companion breadcrumb cover the hierachy..

I also have a particular aversion to these shown/load more transient effects.. as navigation history doesn't preserve I ended up showing all the children with next/previous clicks... perhaps we also need a show all children (like a standard tree expand ;-) )?

Is it not also the case that an author/editor would never use the tree to find content in a blog section (or any node intended to have lots of children).. they'd be using the search on the collection view.. so at that point the show prev/next are redundant?

Maybe we could have an explicit show in tree button? (like in visual studio/rider etc...)

**Comment by MichaelNielsenDK** [comment]
Hi @nielslyngsoe 

Thanks for the feedback. 👍🏻

I agree that seeing the active path in the tree, in some or even many cases, can be useful context, even though we also have the breadcrumb.

But I think the proposed "active path only" solution risks becoming a slightly strange middle ground. If I only see one node from each level, I could still wonder where the other children went. 

The current `...` makes that even worse because, especially with nested Collections, it genuinely looks like something failed to load.

I think there are two valid use cases here, and ideally the Collection/List View could support both:

**1. Allow Collection items to be hidden from the content tree completely.**

   In my example I would simply see the tree down to `Afrika`, and nothing below it. I am completely fine with that, and it is also how this project behaved before the upgrade. Navigation below that point happens through the Collection itself and the breadcrumb still shows where you are.

**2. If items are shown in the tree, make the limited view clearer and more useful.**

   If you want to show nodes in the tree, to provide further guidance to the editor, you should be able to do that. 

   The problem for me is how it currently looks. When only a few nodes are shown and the rest are represented by a bare ..., especially with multiple nested Collections, it very quickly looks like something is missing or has failed to load.

   I would keep the existing idea of showing a limited window of nodes around the active path, but make the controls explicit. For example, use a smaller/subtle “Show previous …” and “Show more …” instead of just ....

   As a secondary improvement, it would also be useful if the number of nodes shown at each Collection level was configurable. One project might work well with 5, another with 10 or 20.
   
   It would look something like this
   <img width="1086" height="1448" alt="list-view-content-tree" src="https://github.com/user-attachments/assets/d0dc2d27-530a-4800-960f-c6bb48af72d5" />
   Of course the show previous and show more wont be there if it's the first or last few items in the collection, or if there arent more than what is set as limit. My example is pure "worst case" scenario 😅 

   And something along the lines of these two settings on the Collection:

   <img width="1790" height="879" alt="list-view-settings" src="https://github.com/user-attachments/assets/c9d4f698-d808-47c4-a602-03b1c9971b9a" />

   If the first setting is disabled, the second one could simply be hidden/disabled.


I think that gives us the best of both worlds: projects where the Collection should effectively replace the tree can keep it completely clean, while projects where the tree context is useful can still have it without potentially rendering hundreds of nodes.

**Reply by nielslyngsoe** [reply]
Hi @MichaelNielsenDK 

Thanks for elaborating on where you stand currently and how you see a potential solution being shaped — a very useful way to capture your perspective.

It becomes clearer that showing other Children of a Collection was wrong. A strong point that came forth earlier was that having many items, like thousands. In that case, seeing the next 10 or 100 would not be useful; it is just wasting the user's screen space.

This leads to it being problematic that we show a single item above and below, problematic because it shows some content, but so little that users get concerned about where the other content is.

Navigating to other items via the Tree for Children of a Collection is not ideal. How the content is sorted in the tree isn't a fit for Content of a Collection — Users would experience a disconnection between the sorting or filter they saw in the Collection.

As well, in the case of Collections inside Collections, having other Children ruins the overview that it was intended to provide — especially clear in your example.

So many things leads towards seeing other Children of a Collection in the Tree is not a fit.
— The only reason why a few were displayed was to give the visual context, so I'm looking for a solution where that is not necessary.

This means the direction is a condensed solution, one that shows only where you're at.

I'll be back with more next week.

**Reply by MichaelNielsenDK** [reply]
Hi @nielslyngsoe

If you have 1000's of items, I would absolutely choose to hide them completely. I agree that showing the next 10, 20 or 100 doesn't really add anything there.

And I agree about your point about sorting/filtering. If the collection is the interface for finding and navigating items, then having the tree present a different representation of the same content can be confusing.

The reason I suggested making it configurable was to leave room for both behaviours. Collection items are currently shown in the tree, and removing them completely again might also cause issues for people who have come to rely on that behaviour.

So to me, being able to choose between hiding collection items completely or showing a limited number still seems like a useful middle ground.

Looking forward to seeing what you come up with 🙂
